### ISSUE TITLE
Caret positioning with initial masked character

### ISSUE THREAD

**Issue by philayres** [igorescobar/jQuery-Mask-Plugin#634]
### Have you take a look into our docs?
https://igorescobar.github.io/jQuery-Mask-Plugin/

### Make sure your read this before opening a new issue:
https://github.com/igorescobar/jQuery-Mask-Plugin#problems-or-questions

#### Device

Ubuntu on multiple PCs

#### Browser (and version)?

Current version of Firefox and Chrome

#### Functional `jsfiddle` exemplifying your problem: 

Use the US Telephone example on https://igorescobar.github.io/jQuery-Mask-Plugin/

When typing a phone number rapidly, as a professional data entry person would do, the caret does not remain at the end of the entry. This only appears to happen when there is a masked character at the start of the field, such as '(' in a US telephone number.

To simulate, press two keys almost simultaneously and almost every time it fails.

#### Describe de problem depth:

This leads to issues for people doing data entry, since it is too easy to accidentally enter incorrect values.

The issue appears to lie in the `behaviour` function, 

            behaviour: function(e) {
                e = e || window.event;
                p.invalid = [];

                var keyCode = el.data('mask-keycode');

                if ($.inArray(keyCode, jMask.byPassKeys) === -1) {
                    var newVal   = p.getMasked(),
                        caretPos = p.getCaret();

                    // this is a compensation to devices/browsers that don't compensate
                    // caret positioning the right way
                    setTimeout(function() {
                      p.setCaret(p.calculateCaretPosition());
                    }, 10);

                    p.val(newVal);
                    /* Do not set the caret position here, since it incorrectly records 
                       the position of the caret before the timeout callback */
                    // p.setCaret(caretPos);
                    return p.callbacks(e);
                }
            },

I experimented with reducing the 10ms timeout to 1ms, and although that helped, I could still occasionally produce the error. Instead, commenting out the setCaret() call where marked in my code above appears to work for me. My guess is that this setCaret() call is not needed, and in fact records the caret position incorrectly before the timeout fires. This leads the setCaret() inside the timeout callback to believe that the caret had been repositioned and therefore leave it in the incorrect position.

Since I'm not familiar with the historical context to why this code is in place, and which devices or browsers are affected, I'll offer this as a solution for my specific issue. Perhaps others can test.

My commit is here: https://github.com/philayres/jQuery-Mask-Plugin/commit/5fc2cd5723acab64ef214f4089a88747766d8708

Happy to submit a pull request if you agree that this is an issue and a reasonable resolution.

**Comment by KITSDmytroKovalenko** [comment]
I've got the same issue using automated software. That's really critical from test automation point of view.

**Comment by philayres** [comment]
@KITSDmytroKovalenko I found that with my automated tests I needed to send a sequence of individual keypresses separated by 20ms sleeps to get the entry to work. The commit I suggested above appeared to resolve this for me.

**Comment by igorescobar** [comment]
It's settled to 20ms for a reason. Maybe I could turn this into a global
variable and then you can set to whatever you guys need without compromise
stability to everybody else.

What you guys think?

**Comment by philayres** [comment]
I don't think that adjusting the timeout really fixes the actual problem. At least according to my project's tests, it can be fixed completely, as I say in my original post: 

                p.val(newVal);
                /* Do not set the caret position here, since it incorrectly records 
                   the position of the caret before the timeout callback */
                // p.setCaret(caretPos);
                return p.callbacks(e);

I now don't see an issue, no matter how fast I enter characters.

**Comment by igorescobar** [comment]
I've just released a new version, can you please check it out? @philayres @KITSDmytroKovalenko ?

**Comment by philayres** [comment]
@igorescobar I can see from the US phone number test on https://igorescobar.github.io/jQuery-Mask-Plugin/ that this has not been resolved. Did my suggestion in the original post not work, or cause other issues? I see from the commit that you made https://github.com/igorescobar/jQuery-Mask-Plugin/commit/62ca6f68ee757f3e54d7af9189db4dca9b9c2c8d#diff-22fc0e143f509f85a43bc863e773f0cd that you only provided a way to adjust the time. 

I still believe that the root cause of the issue is the unnecessary call to `p.setCaret(caretPos);` in the `behaviour` function, just before the `return`. I have removed that in the code on my test and production systems for some time with no issues.

I would suggest that you re-open this issue, since the current version has not resolved the problem.

**Comment by igorescobar** [comment]
@philayres It's open :-)

I understand that you think it's unnecessary call but it isn't. Removing it brakes a lot more than it fixes :)

**Comment by philayres** [comment]
OK. So from what I can tell, perhaps that call needs a check for keypresses that don't result in a character (such as backspace, cursor keys, etc). I see that some things are broken in my version when navigating without this, but at least I can enter data correctly. 

Would this correspond with the functionality you think is broken when when removing the call I suggested? If so, I'll try a new version of the code to see if it helps.
